### DISCUSSION TITLE
Discussion of usage in a Qt application

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#716] [General]
As @vkrause pointed out in https://github.com/nu-book/zxing-cpp/issues/86#issuecomment-641393971, there are a couple of KDE apps using this library. I had a look at the three code bases and have a few comments / suggestions:

Two of the apps only deal with decoding QRCodes and share the following issues
 * they set `tryRotate()`, even that has no effect on the QRCode decoder
 * they could set `tryHarder(true)`, which should be considered the default (see also #104)
 * they create a `GenericLuminanceSource` with the RGB constructor overload but set the `pixelBytes` parameter to 1, which is a bug in any case (is this particular QImage guaranteed to contain only grey-scale bits?)
 * they could both take advantage of the 'new' `ReadBarcode` function and simplify the code that way or alternatively go one level deeper and use `ZXing::QRCode::Reader`
 * they could use `QString::fromWCharArray` (see [here](https://stackoverflow.com/questions/14726304/best-way-to-convert-stdwstring-to-qstring)) instead of using the `TextUtfEncoding`

The other deals with scanning several 2D codes. It provides it's own `BinaryImage` implementation as a wrapper for `QImage` but would effectively only work properly with 'ideal' black/white input. Why is that code not using the ZXing binarizer code? Performance?

Also related to the discussion in #75, I'd like to start some trivial Qt support basically offering a `ReadBarcode(const QImage&, ...)` function. This can be implemented in a header-only way (not introducing any Qt build dependency) and prevent most of the issues above.

**Comment by axxel** [comment]
Done: 
 * first trivial Qt wrapper in form of an [example application](https://github.com/nu-book/zxing-cpp/blob/master/example/ZXingQtReader.cpp) to have a look at
 * deprecated the `isPureBarcode` function (is never called anymore)
 * fixed the regression that caused @vkrause's pure-code-path to crash at some point in time (i hope)

ToDo:
 * automatic quite zone extension/insertion for the 'pure' barcodes

Feedback welcome. Also from @swex maybe?

**Comment by axxel** [comment]
I just added a new `ZXingQtCamReader` example that is something like a stripped down version of what @swex QZingNu wrapper does. The noticeable difference is that it does away with all the `QVideoFrame` to `QImage` conversion and uses the video frame data directly. That removed two image copies and all pixel conversions from YUV -> RGB -> Grey. The code is simpler, faster and also produces no additional noise during the pixel format conversion. Disclaimer: most of the 'supported' pixel formats from QVideoFrame are not tested, due to lack of input.

There are now overloads for `ReadBarcode` that take a `QImage` or a `QVideoFrame`. Additionally, there is a video filter that can be used in QML applications.

What is still pretty much a work in progress is the right way to export the public ZXing enums, like `BarcodeFormat` without repeating their definition and introducing a second, incompatible enum class.

My intention would be 'officially' include the `ZXingQtReader.h` as a header only access to the wrapper without adding a separate library linked to qt to the project. This will require a little more care with the moc generation when incorporating that code to a client project. I'd appreciate any feedback and help related to best practices regarding the moc/meta-data setup.

**Comment by axxel** [comment]
Very good to hear. So all your samples work with current HEAD in `isPure` mode?

Regarding the flipped PDF417, in case you have not noticed, thanks to @tailsu you don't need to deal with flipping images outside of ZXing anymore. My upcoming `pure` special casing for PDF417 will keep that feature and will additionally spare you from trying rotated symbols as well. With non-square symbols like PDF417 I see why people would choose to render them rotated onto their PDFs.

Anything else necessary for you to switch to `ReadBarcode()`?

**Comment by vkrause** [comment]
I've committed a code path using ReadBarcode now (and removed the usage of other deprecated API we still had in there while at it), conditional on a version number greater than 1.1.1, as we rely on all the recent fixes for this. So the next ZXing release would automatically complete the transition for us.

I'll revisit our hacks for rotations/flips, this predates a lot of improvements in ZXing as well as fixes for shortcomings in our PDF handling code, it is quite possible that parts or even all of that are no longer needed in this form.

**Comment by axxel** [comment]
Nice. A couple things I'd like to comment on:

* In the `ZXING_USE_READBARCODE` code path you only need to include `ZXing/ReadBarcode.h`
* The other two projects Qrca and Kaidan would need to be ported as well. Since they are only concerned about non-pure QRCodes and apparently only get Greyscale8 input, they can in fact use release 1.1.0 with its `ReadBarcode` implementation. Are you in a position to do that or do I need to contact someone else?
* Do you see any value in or have input for my proposed 'upstream' `ZXingQtReader.h` wrapper?

**Comment by axxel** [comment]
Addendum:
 * With my latest PDF417 improvements you should be able to drop all your rotation/flip special casing.
 * The new isPure approach should be considerably faster overall compared to the 1.1.1 code path. I'd be curious to get some feedback from you whether you can confirm that. I'd argue that your code that selects the decoders based on aspect ratio is not required anymore either.
 * Building for an architecture that support SSE or even AVX will make a difference, too.

**Comment by vkrause** [comment]
With the latest version I see a number of regressions with PDF417 decoding unfortunately, I'll email you a few samples.

Regarding the Qt wrapper, I don't see a way of providing Qt introspection features with a header-only setup. However the most valuable thing in there is the QImage -> ImageView adapter code IMHO, which should work header-only.

**Comment by Thaodan** [comment]
Talking about Qt integration: would it make sense to add a QML Plugin?

**Comment by axxel** [comment]
@Thaodan Have you seen [ZXingQtCamReader.cpp](https://github.com/nu-book/zxing-cpp/blob/master/example/ZXingQtCamReader.cpp)? Is there anything missing to make `ZXingQtReader.h` it a 'QML Plugin'?

**Comment by Thaodan** [comment]
> @Thaodan Have you seen
> [ZXingQtCamReader.cpp](https://github.com/nu-book/zxing-cpp/blob/master/exa
> mple/ZXingQtCamReader.cpp)? Is there anything missing to make
> `ZXingQtReader.h` it a 'QML Plugin'?
> 
No I don't but was thinking more of something that is in a form of library as 
a dynamic plugin so it can be shipped along the main library.
But this looks like a good start.


Also I might add QZXing already had better Qt integration in QImage and such.
This could be a good example to start up on in addition.

**Comment by axxel** [comment]
> Also I might add QZXing already had better Qt integration in QImage and such.

What do you believe is missing?

**Comment by kosadev** [comment]
@axxel Do you think that it would be beneficial to put the QR scanner thing in the Qt repo? That would be definetely a good thing for the community, but I don't know how Qt maintainers will see that.

**Comment by axxel** [comment]
@kosadev I don't think I want to get the Qt company involved in this. My idea was rather to add a 'Qt wrapper' to this repository to prevent even more fragmentation: @swex started something similar to your effort 4 years ago but has seemingly abandoned it. Then there is a sizable community around https://github.com/ftylitak/qzxing. I got in touch with @ftylitak years ago about merging efforts but he was heavily invested in backward compatibility and did not want to commit to a c++14 build dependency. We are at c++17 now and I'm eyeing c++20 already, so there is likely not much point in bringing this up again (even though I am just doing it right now ;)).

The question is: do you see value in (partly) upstreaming your functionality? I did not have a close look at it, yet. It might be too specific to your own needs for me to consider it fitting. As you can see above and in the sample code, my idea was to have a very light header only wrapper that is trivial for Qt people to use. It got stuck there as an example because of lack of interest. I'm not against making it into a library/plugin if there is a technically convincing argument.

**Comment by vkrause** [comment]
Btw regarding generic Qt/ZXing integration, on the KDE side we meanwhile also have https://invent.kde.org/frameworks/prison/-/tree/master/src/scanner for scanning live video feeds from Qt Multimedia 5/6 (QML usage examples can be found further up in the same repo). That doesn't cover working with a single image though.

**Comment by axxel** [comment]
@vkrause Interesting! So now there is yet another implementation of basically the same functionality. Only this time even including Qt6 support. So same question to you: would there be any value for you if something like your own version would be inside this repo? Can you / @kosadev quickly asses whether your two versions are in fact essentially the same (apart from the obviously missing `QImage` and the writer functionality that is present in https://github.com/scytheStudio/SCodes?).

Side question: are you really sure that `QVideoFrame` objects never have a row-stride other than the width x pixel-size of the frame? I would have expected this to be heavily dependent on the specific video capture hardware involved.

**Comment by vkrause** [comment]
Interesting indeed, I didn't know the SCodes implementation yet. One difference I can spot on a quick look is that this seems build around QString for the barcode content. This turned out to be problematic for us when dealing with binary rather than textual contents. That's why this is a bit more complex in our API, supporting both QString and QByteArray.

For barcode generation we actually have a dedicated library next to this as well, which is much older already and predates our use of ZXing (and thus uses other libraries and its own code to generate most formats). This is something I'd personally like to see phased out in favor of ZXing, but this isn't entirely up to me :)

Regarding the video frame formats, this should be the case for the formats we handle explicitly for zero-copy processing by ZXing, other formats that are either not fully specified or that cannot be directly consumed go through a QImage conversion first. That's provided by Qt, which hopefully knows how to deal with that specific format (it has to anyway to eventually render this on the screen).

**Comment by axxel** [comment]
> QString / QByteArray

Sure. You are not the first to have to deal with the issues arising during a `std::wstring` to native string object conversion when binary data is involved: @markusfisch had to come up with a hack in the android/JNI context as well. What I'm saying is: looks like a generic problem that belongs into such a wrapper in some shape or form.

> barcode generation

see new issue #332 for further discussion.

> video frame formats / row stride

I was not talking about the generation of `QImage` objects from `QVideoFrame` but the interpretation of `QVideoFrame` data in your zero-copy code. I strongly believe that you have a nasty bug there that will show up sooner or later, once someone gets there hands on some video-capture hardware that returns their frames with a row-stride other than `width x pixelSize` and I'm pretty sure they exist. That is the reason for the existence of the `bytesPerLine()` member.

> 'upstreaming' your code

Interestingly, you did not address my first question. I interpret that as a hint that the answer is closer to no than to yes. But I'd prefer to hear your opinion, instead of speculating about it ;).

**Comment by vkrause** [comment]
> video frame formats / row stride

Ah, I see what you mean. https://invent.kde.org/frameworks/prison/-/merge_requests/22 should address that.

> upstreaming

Sorry, not answering that part of the question was just me being sloppy, not in any way intentional :) I would indeed see value in a joined implementation over maintaining our own, as you say feature-wise this is all very close together anyway, so it's extremely likely that we'd be able to come up with something that matches our needs. Where exactly that is located is secondary IMHO, the main point for me is that it saves all of us some work.

Our current implementation is the result of unifying a few application-specific implementations we had of this, doing that inside our infrastructure was just the easiest and fastest way, and I admittedly didn't think about doing this in an even wider context.

**Comment by ApiTracer** [comment]
Hello,

I'm facing all issues raised by @axxel, so I want to contribute to the Qt6 port and the concentration of efforts to provide a great Qt port of zxing.
Some contexts of my use :
- I use zxing in my Android/IOS Qt app 
- The use is outdoor detection of small 1D code-128 barcodes. Some screenshots :
![scanner_screenshot](https://user-images.githubusercontent.com/74172673/171276846-5e010678-1e78-40b3-bc84-2949d211350d.png) (real screenshot)
![barcode_label](https://user-images.githubusercontent.com/74172673/171276512-c3a555a2-3d94-4cfe-8da5-d51b30892a71.png) (new label form, but not real image)
- Qt5 app and zxing integration works pretty well. I tested several integrations (zbar and severals zxing integrations), to finish with Scodes integration for the speed, in part due to this zxing-cpp. At qt5 time, goals was improve speed on old Android phone and iPhone, avoid lag after long scan period (typical use is few minutes active scanning to find several different barcodes : ~100), and decrease battery consumption.
- I needed to port my app to Qt6, in particular for billing integration of Android. Problems started with that.
- I start by trying to port Scodes to Qt6, then I try others integrations (Prison of KDE and others) and finish to find QZXing Qt6 branch. It works on IOS but slow and crash on Android. Thanks to this Qt-Bug(https://bugreports.qt.io/browse/QTBUG-97789), I can make works the QZXing on Android and IOS but it's too much slow compare to Qt5 Scodes, so I can't update my app.

I begin to try a Qt6 port of the Qt example of this repo, but it will be my first contribute to GitHub public project and I don't known well header-only. @axxel are you ok to help me to start the port in the right way and to help to make a general example if I success to make a working but sub-optimal project ?

**Comment by axxel** [comment]
@ApiTracer Thanks for joining the discussion. You wrote that you tried Prison of KDE but did not end up using it. May I ask why?

**Comment by ApiTracer** [comment]
The main reason, it's I haven't success to config all the KDE workspace, especially Extra-make-module. brew install of ECM works with qt creator after some settings but it raise errors. And it seems that it's not works with IOS.

**Comment by axxel** [comment]
I see. I would suggest to do the following: use the Prison code as an inspiration/template to upgrade the current `ZXingQtReader.h` for Qt6 support. If that works, we can then decide on how to move forward regarding the question of whether or not to make a Qt lib out of that.

**Comment by ApiTracer** [comment]
Hello,
I have a first try : https://github.com/ApiTracer/ZXingQtCamReader
It works on Android but it's very slow. The scanner make lags on UI thread.
On IOS, it's not works for now because the lib isn't linked in static. I have already fix it on another project, I will need to look it after. 

I tryed to avoid this lags, I think too much frames is scanned. But it crash when I add boolean or Timer to process only some frames. I think is linked to multithreading. 

Screenshot that show that the project works with the android simulator on macOS. but the TV of the android simulator show there are very few fps.
<img width="1002" alt="Capture d’écran 2022-06-02 à 15 19 51" src="https://user-images.githubusercontent.com/74172673/171655074-1bcfe926-cc20-4609-902a-fbf74030d018.png">

![Screenshot_1654175995](https://user-images.githubusercontent.com/74172673/171655437-088a1065-0d2d-42ea-9a9a-ecea67ed3420.png)

**Comment by axxel** [comment]
@ApiTracer thanks for sharing. I can't really evaluate the changes you made, as you copied the code into a new project, so I have no diff. If you want to pursue this further, please clone this repo and make your changes there. I saw that you simply modified the code to work with qt6. Please take another look at the KDE code and make similarly sure that it compiles with both qt5 and qt6.

**Comment by emericg** [comment]
I have made a demo application for Qt6 and barcode scanning.

https://github.com/emericg/QmlMobileScanner

Since Qt6 and the new Qt Multimedia module, the camera feature (and everything else unfortunately) is mostly broken, so success will depend a lot on what version is used (tested successfully with 6.5.3 and 6.6.1) and also what backend is chosen at run time (the new new ffmpeg backend for Qt Multimedia is at this time even more broken than the old backends).
This app forces the old android backend for instance, because the ffmpeg one cannot pass frames through video filters anymore.
There are also various Qt crashes when you stop cameras, or minimize apps and bring them back into focus... Hopefully everything will be squared away with time.

- The "official" qzxing project is mostly broken on Qt6, so the backend included in QmlMobileScanner has multiple fixes and improvements, and now run well with Android and iOS.

- The "official" zxingcpp Qt demo is a bit limited, so I've made a proper backend for the needs of this demo application, which adds efficient scanning using QtConcurrent and an image provider to generate barcodes with ease.
It's still about 2 to 3 times slower than qzxing, but it can scan a lot more formats obviously!

- The zint backend is also new, also with an image provider to generate very exotic barcodes with simple integration into Qml applications. It's a toy more than a proper backend, but it is quite fun to generate GridMatrix or UltraCodes :)

**Reply by axxel** [reply]
Thanks for reviving this old and still unsettled discussion and thanks for sharing your project. Here are my thoughts:
* I understand that qzxing is still around because it has a sizable user base who rely on the API as is. But for someone writing some completely new code actually abstracting away that API (like you did), I don't see the value in having those two backends. Can you give some motivation for that?
* Your statement `It's still about 2 to 3 times slower than qzxing` surprised me quite a bit, I have to say. After all the work I spent over the last years on improving its performance, I can claim quite confidently that zxing-cpp is the fastest open source barcode scanning library on the planet right now. I must therefore conclude that you either do something 'wrong' or you are comparing apples to oranges, so to speak. So I obviously had to see for myself how current qzxing actually fares against current zxing-cpp. I have to say, on a first glance it is indeed faster than I expected. The biggest issue in comparing the two is actually the apples vs. oranges situation. E.g. qzxing generally only looks for a single barcode and I just added an option to `ZXingReader` to do the same for easier comparison. Then zxing-cpp can detect a lot more difficult codes than qzxing, especially Aztec, DataMatrix and QRCode. And it can detect completely new formats. You mentioned that yourself. Anyway, to get some meaningful comparison, I decided to take the most standard use case of detecting only QRCodes. In terms of speed in an interactive environment, it is most important how long the library takes for looking at an image without any barcodes in it. So compared `ZXingReader -format qrcode -noscale -single -binarizer global` with the simple qzxing test code that follows below and had both look at one of the falsepositive samples and the results where (drumm roll):
qzxing takes 6.7ms while zxing-cpp takes 0.6ms. So if you are happy with the reduced detection rate caused by those settings (and that is what you roughly get when using qzxing), then zxing-cpp can provide you that **10x faster** and if I let both look for all supported formats, zxing-cpp is still 2x faster.
* If you have zint as a backend, I'd focus on that. Its quality and breadth *surely* outperforms what qzxing or zxing-cpp have to offer on the generating side (see also #332).

```c++
#include "QZXing.h"

#include <iostream>

int main(int arc, char** argv)
{
    QImage img(argv[1]);

    QZXing decoder;
    decoder.setSourceFilterType(QZXing::SourceFilter_ImageNormal | QZXing::SourceFilter_ImageInverted);
    decoder.setTryHarderBehaviour(QZXing::TryHarderBehaviour_ThoroughScanning | QZXing::TryHarderBehaviour_Rotate);
    decoder.setDecoder(QZXing::DecoderFormat_QR_CODE);

    QElapsedTimer timer;
    timer.start();

    QString result;
    int i = 0;
    do {
        result = decoder.decodeImage(img);
        ++i;
    } while (timer.elapsed() < 1000);

    std::cout << result.toStdString() << " (in " << (float)timer.elapsed() / i << "ms)\n";

    return 0;
}

```

**Reply by axxel** [reply]
Little update on the performance situation: with the latest changes in HEAD and using gcc instead of clang the above mentioned `ZXingReader` call measures only 0.44ms, i.e. zxing-cpp is 15x faster than qzxing in that particular scenario.

**Reply by emericg** [reply]
> Can you give some motivation for that?

This application is more intended as a demo and an example for Qt developers. Having both QZxing and zxing-cpp has been beneficial to test both in the same environment while working on the backends. 
Other developers have already built the app and we can compare how cameras works on various phones and OS, where are the bugs in Qt Multimedia...

I also like to think that the QZxing version integrated in the app can be the "definitive edition", because it just works with Qt 6.5+ on every platform, which is not the case at all with the upstream repository (and it doesn't merge legitimate bug fixes anymore, it might be the end for QZXing).
I have older applications that have a QZxing integration, and they can still use this code as is. I probably won't use it in future developments.

> Your statement It's still about 2 to 3 times slower than qzxing surprised me quite a bit, I have to say

I meant no disrespect to your work, and you are right this is probably an apple to oranges comparison. I can say that zxing-cpp is way better at detecting qrcodes (and others) in challenging situations. I also really like that it can easily generate more than just qrcodes.

But still, the fact remains that with similarly built Qt backends, using data from the camera, it is about twice as slow.
I double-checked if something obvious was different between the two, and I tried disabling everything but QrCodes, and disabling the various tryHard, rotate, setting a single code detection... and I am not anywhere close to 10-15x faster than QZxing.
I also checked the backend from the kde developers, I don't see anything fundamentally different between our codes.

I'm using a recent Android phone to test, so that means llvm and a weak cpu. (The application works on computers, but detecting barcodes on fixed webcams isn't really practical, nor a target use case for me.)
The run time is about 60ms per frame (for the simplified one qrcode test), each QVideoFrameFormat(Format_RGBA8888, QSize(2160, 3840)) is processed in a thread, and I don't process multiple frames at once. I can see that half of that seems to be mapping the webcam gpu buffers to cpu memory before handing them to zxingcpp, but that part is the same between both backends.

I would call this "good enough" to be used in a mobile barcode scanner, but we can spend some time to try and understand why the speed is what it is if you have some time to explain what would be the best path to success.

- We can already adjust tryHarder, tryRotate, tryInvert, tryDownscale, the capture area, and the barcode formats to scan for dynamically in the app UI? Is there something else I should consider for performance?
- When the camera frames are mapped to CPU memory and copied, we can specify what would be the optimal pixel format for zxing-cpp?
- After that, I crop the buffer manually to remove about half the pixels (optional step), so that mean another copy. It seems that both qzxing and prison are doing the same. I tried using ZXing::ImageView.cropped() but I found it slower than doing the crop myself?
- I tried scaling the picture size myself but found it counterproductive regarding detection rates.
- I could have one (or two) thread producing memory buffers from the gpu frames, and another (or two) consuming these and doing the detection, we would have more detections per second, but that would not explain the performance delta between the two backends anyway.

> If you have zint as a backend, I'd focus on that

Again I'm not about to make the best barcode scanning and generating application on the Internet, I just did this as a demo.
Zint is a complex piece of software, and for simple use cases, I strongly think using zxing-cpp is just much easier. I understand you don't want to extend that part, but people should not have to use zint to display an EAN 13 barcode on screen...

**Reply by axxel** [reply]
> This application is more intended as a demo and an example for Qt developers. [...]

Thanks for the explanation. 

> ... (and it doesn't merge legitimate bug fixes anymore, it might be the end for QZXing).

I came to the same conclusion, let's see if I get a reaction to [this comment](https://github.com/zxing-cpp/zxing-cpp/issues/75#issuecomment-1933232670).

> I meant no disrespect to your work

I did not see any :). I was just surprised and of course it bugged me, since I'd like people to see the benefits of my work (otherwise it would have been for nothing because I personally have indeed no use for a barcode scanning library ;)). Still thinking about the lesson for me to learn here. See also below.

> I'm using a recent Android phone to test, so that means llvm and a weak cpu. The run time is about 60ms per frame (for the simplified one qrcode test), each QVideoFrameFormat(Format_RGBA8888, QSize(2160, 3840)) is processed in a thread, and I don't process multiple frames at once. I can see that half of that seems to be mapping the webcam gpu buffers to cpu memory before handing them to zxingcpp, but that part is the same between both backends.

Two things:
 1) you are saying that copying your 8MP image from the GPU to the CPU takes 30ms?
 2) 8MP is indeed rather large for a hand-held, interactive use case.
   2a) QZXing seems to forcefully scale everything down to under 1MP before even looking at the data.
   2b) the zxing-cpp demo app for Android works on 1280x720 on my Pixel5 and Pixel3 and takes about 10ms (and
        there is something fishy with that number because it hardly changes between QRCode-only and Any-barcode setting).

If you can build and run the Android demo-app on your system, Id be very interested to learn how it compares to your own Qt based app.

> I would call this "good enough" to be used in a mobile barcode scanner, but we can spend some time to try and understand why the speed is what it is if you have some time to explain what would be the best path to success.

I'm certainly interested in this. Thanks for the offer.
 
>  We can already adjust tryHarder, tryRotate, tryInvert, tryDownscale, the capture area, and the barcode formats to scan for dynamically in the app UI? Is there something else I should consider for performance?

For the qzxing vs zxing-cpp comparison, you'd need to set the binarizer to `GlobalHistogram` as that one is [hardcoded](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/QZXing/QZXing.cpp#L393) in qzxing. Also the most likely most relevant difference is the above mentioned automatic [downscaling](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/QZXing/QZXing.cpp#L370). If zxing-cpp is half as fast with 8MP input as qzxing is with 1MP, then making them work on the same resolution should get you a lot closer to my claimed 10x difference.

>  When the camera frames are mapped to CPU memory and copied, we can specify what would be the optimal pixel format for zxing-cpp?

Definitely grayscale. The fastest possible route from the image sensor to a barcode would be as follows:
 1. request a "reasonably sized" frame from the camera. I'd argue that 8MP is overkill for an interactive application where the user points the device at a barcode.  720p or 1080p seem reasonable to me.
 2. request a "native" pixel format that is ideally not RGB but some Yuv variant (like any of [those](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/zxing-cpp/wrappers/qt/ZXingCpp.cpp#L188)).
 3. use the `ReadBarcodes(QVideoFrame, ...)` overload I implemented to get direct access to the pixel data without first creating a QImage.

If you can't specify the frame buffer size to something reasonable, then going through QImage and do a manual resize will be faster than processing the full frame. In that case, first convert to grayscale, then resize this should be faster.

These best practices are not obvious to the average GUI developer. I can only conclude that I need to make sure zxing-cpp comes with a tuned and usable Qt wrapper right out of the box. If you have insights as to what you found lacking from the header-only `ZXingQtReader.h` approach, please let me know.

>  After that, I crop the buffer manually to remove about half the pixels (optional step), so that mean another copy. It seems that both qzxing and prison are doing the same.

I have not seen any internal default cropping in qzxing but maybe I've overlooked this.

> I tried using ZXing::ImageView.cropped() but I found it slower than doing the crop myself?

That is surprising. No idea why that would be the case. If you crop anyway and need to go through a resized QImage, then crop before doing the resize/rgb2gray operations.

>  I tried scaling the picture size myself but found it counterproductive regarding detection rates.

I see, but if you want to compare apples to apples, you'd need to pass the commented width and height parameters [here](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/QZXing/QZXingFilterVideoSink.cpp#L91) or scale down the zxing-cpp input to `< 999x999`.

>  I could have one (or two) thread producing memory buffers from the gpu frames, and another (or two) consuming these and doing the detection, we would have more detections per second, but that would not explain the performance delta between the two backends anyway.

I don't see any reason to resort to multiple threads working on different frames in parallel. What does make sense is to have a thread separate from the main GUI thread, so the UI does not get blocked.

> Zint is a complex piece of software, and for simple use cases, I strongly think using zxing-cpp is just much easier. 

I see. My intention is to use zint as the backend for the generation via a simple `WriteBarcode` function, analogous to the `ReadBarcode` one.

**Reply by emericg** [reply]
> 8MP is indeed rather large for a hand-held, interactive use case.

So I've worked out how to change the camera settings, and I can choose from a variety of resolutions, each with either YV12 or NV21 pixel format. It's not very practical, and I'm not entirely sure that it is really respected for all modes.

Also the camera may capture NV21 buffers, but the one we get in with QtMultimedia through QVideoFrame are always RGBA8888 in video memory. It makes a lot of sense for most use cases, unfortunately for ours, we lose that "perfect" luminance channel.

> you are saying that copying your 8MP image from the GPU to the CPU takes 30ms?

Yes, it's certainly not good. And probably buggy...
Fun fact, if I set the same resolution for both front and back camera (for instance 1200x1200, the front camera can't go as high as the other) I have like 45 ms per frame for the back camera and 20 ms for the front one...

> use the ReadBarcodes(QVideoFrame, ...) overload

That function was a lot slower, but there was a missing case for RGBA8888 and so it was falling back to the other function. That being said now it's only a bit slower...

> If you can't specify the frame buffer size to something reasonable, then going through QImage and do a manual resize will be faster than processing the full frame. In that case, first convert to grayscale, then resize this should be faster.

Same thing, it's hard to get performance improvements that way. Scaling or converting the buffer to greyscale always leads to slightly worst performances :/ I'll keep looking at what's going on.

> For the qzxing vs zxing-cpp comparison, you'd need to set the binarizer to GlobalHistogram as that one is [hardcoded](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/QZXing/QZXing.cpp#L393) in qzxing. Also the most likely most relevant difference is the above mentioned automatic [downscaling](https://github.com/emericg/QmlMobileScanner/blob/2aeccab3bb22e9d9f9dfc8a3c3c1cab1b2e968a5/src/thirdparty/QZXing/QZXing.cpp#L370).

I can confirm that if I set QrCode only and the same binarizer, and remove the downscale for qzxing, the results are now real close.

> I have not seen any internal default cropping in qzxing but maybe I've overlooked this.

This is just the UI choosing to scan only inside a rectangle on the screen.
QZXingFilter::processFrame() simply does something like that frameToProcess = image.copy(captureRect);

> If you can build and run the Android demo-app on your system, Id be very interested to learn how it compares to your own Qt based app.

I'll have a look at this next week, it should be interesting. But I don't have much experience with pure Android app.

All in all there are still a lot of things I have to try, and benchmark everything a little bit more seriously.

**Comment by axxel** [comment]
In case someone wonders: I converted this old issue to a discussion to make it easier to follow different strands. (I believe the discussion feature has not been available yet when I open the original issue ;)). Then I removed a whole bunch of entries that were about nothing but fixing random bugs and had no real value for the general "usage in a Qt application" subject.

**Comment by axxel** [comment]
I'll comment on a Qt examples specific subject from #690 here, because it is very much related to the general Qt usage question:

The single-barcode reading limit of the Qt example is just a relic of the past, basically. I did not change that yet for 3 reasons:
1) it is unclear to me whether it would be more meaningful to have the `BarcodeReader` emit one signal per image that returns a list of barcodes or whether it should rather emit multiple events in that case. Then there would also need to be a 'startOfFrame' signal. which might be weird.
2) my QML skills are definitely not up to the task of displaying a list of barcodes.
3) there seemed to be no interest in the Qt code anyway.

**Reply by axxel** [reply]
@emericg If you had some of your QML knowledge to spare and could provide a PR that adds support to display/highlight multiple barcodes to the [ZXingQtCamReader](https://github.com/zxing-cpp/zxing-cpp/blob/master/example/ZXingQt6CamReader.qml), that would be splendid ;). It would not need to be fancy (like your color coding thingy), in fact I'd prefer the sample code to stay as short/simple as possible while still showing the libraries capabilities.

**Reply by emericg** [reply]
Sure, I'll have a look at it next week

**Comment by axxel** [comment]
To everyone listening: triggered by the work of @m7913d in #945, I have continued that and significantly overhauled the Qt demo code and put it in it's own [wrapper](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/qt) for better visibility, including a [README](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/qt/README.md).

This new code supports both reading and creating barcodes, returns / emits a vector of all detected barcodes, includes a completely new QWidget based demo app, and has much improved naming consistency in the API. The work is not completed, yet. Any feedback for further improvements are welcome.

One question to the community: Why would I care about still supporting Qt5? (The current code still contains qt5 code but I can't compile / test this anymore, so maybe it is already broken).

**Reply by emericg** [reply]
Hi, that's great new, I'll give it a try when I have some free time.
You should definitely nuke the Qt5 support. It is unusable outside of windows & older linux platforms. And we are actually way closer to Qt7 than we are of Qt5 !
